lightclient/src/background.rs

```diff
@@ -253,129 +357,133 @@ impl<TPlatform: PlatformRef, TChain> BackgroundTask<TPlatform, TChain> {
-                        );
-                    }
-                } else if let Some(subscription_id_state) =
-                    chain_data.id_to_subscription.remove(&id)
-                {
-                    if subscription_id_state
-                        .sub_id_sender
-                        .send(Err(LightClientRpcError::Request(error.to_string())))
-                        .is_err()
-                    {
-                        tracing::warn!(
-                            target: LOG_TARGET,
-                            "Cannot send method response to id {id} chain={chain_id:?}",
-                        );
-                    }
-                }
-            }
             Ok(RpcResponse::Method { id, result }) => {
                 let Ok(id) = id.parse::<usize>() else {
                     tracing::warn!(target: LOG_TARGET, "Cannot send response. Id={id} chain={chain_id:?} is not a valid number");
                     return;
                 };
 
                 // Send the response back.
-                if let Some(sender) = chain_data.requests.remove(&id) {
+                if let Some(sender) = self.requests.remove(&id) {
                     if sender.send(Ok(result)).is_err() {
                         tracing::warn!(
                             target: LOG_TARGET,
                             "Cannot send method response to id={id} chain={chain_id:?}",
                         );
                     }
-                } else if let Some(pending_subscription) = chain_data.id_to_subscription.remove(&id)
-                {
-                    let Ok(sub_id) = result
-                        .get()
-                        .trim_start_matches('"')
-                        .trim_end_matches('"')
-                        .parse::<usize>()
-                    else {
+                } else if let Some(pending_subscription) = self.pending_subscriptions.remove(&id) {
+                    let Ok(sub_id) = serde_json::from_str::<SubscriptionId>(result.get()) else {
                         tracing::warn!(
                             target: LOG_TARGET,
-                            "Subscription id={result} chain={chain_id:?} is not a valid number",
+                            "Subscription id='{result}' chain={chain_id:?} is not a valid string",
                         );
                         return;
                     };
 
                     tracing::trace!(target: LOG_TARGET, "Received subscription id={sub_id} chain={chain_id:?}");
 
-                    let (sub_id_sender, active_subscription) = pending_subscription.into_parts();
-                    if sub_id_sender.send(Ok(result)).is_err() {
+                    let (sub_tx, sub_rx) = mpsc::unbounded_channel();
+
+                    // Send the method response and a channel to receive notifications back.
+                    if pending_subscription
+                        .response_sender
+                        .send(Ok((sub_id.clone(), sub_rx)))
```

Comment:
Could we avoid allocation here by using `Arc<str>` for subscription ids? Probably irrelevant though, feel free to ignore.

Reply:
I'd guess it's not worth while because we (technically, but not really in reality) need to allocate a `String` anyway for the deserializing bit, and then the `Arc` is another alloc, and we onlyclone it once anyways :)